ddtrace/ci_visibility/_recorder.py

```diff
@@ -0,0 +1,143 @@
+import json
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+import ddtrace
+from ddtrace import Tracer
+from ddtrace.ci_visibility._filters import TraceCiVisibilityFilter
+from ddtrace.contrib import trace_utils
+from ddtrace.ext import ci
+from ddtrace.ext import test
+from ddtrace.internal import atexit
+from ddtrace.internal import forksafe
+from ddtrace.internal.compat import parse
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.service import Service
+from ddtrace.settings import IntegrationConfig
+
+
+log = get_logger(__name__)
+
+
+def _extract_repository_name_from_url(repository_url):
+    # type: (str) -> str
+    try:
+        return parse.urlparse(repository_url).path.rstrip(".git").rpartition("/")[-1]
+    except ValueError:
+        # In case of parsing error, default to repository url
+        log.warning("Repository name cannot be parsed from repository_url: %s", repository_url)
+        return repository_url
+
+
+class CIVisibility(Service):
```

Comment:
Can we add tests for the `CIVisibility` service. Specifically test cases that involve forking and updating `CIVisibility._instance` across threads

Reply:
On second thought I removed the `_restart()` implementation for this service, since the CIVisibility isn't meant to be used and restarted on multiple threads/forks. I will add test cases to provide coverage for enabling and disabling the `CIVisibility` service however.